packages/tools/src/tools/annotation/planarFreehandROITool/openContourEditLoop.ts

```diff
@@ -0,0 +1,548 @@
+  const { viewport } = enabledElement;
+
+  const prevCanvasPoints = annotation.data.polyline.map(viewport.worldToCanvas);
+
+  const { spacing, xDir, yDir } = getSpacingAndXYDirections(
+    viewport,
+    this.configuration.subPixelResolution
+  );
+
+  this.commonEditData = {
+    prevCanvasPoints,
+    editCanvasPoints: [canvasPos],
+    startCrossingIndex: undefined,
+    editIndex: 0,
+  };
+
+  this.commonData = {
+    annotation,
+    viewportIdsToRender,
+    spacing,
+    xDir,
+    yDir,
+  };
+
+  state.isInteractingWithTool = true;
+
+  element.addEventListener(
+    Events.MOUSE_UP,
+    this.mouseUpOpenContourEditCallback
+  );
+  element.addEventListener(
+    Events.MOUSE_DRAG,
+    this.mouseDragOpenContourEditCallback
+  );
+  element.addEventListener(
+    Events.MOUSE_CLICK,
+    this.mouseUpOpenContourEditCallback
+  );
+
+  hideElementCursor(element);
+}
+
+/**
+ * Deactivates and cleans up the closed contour edit event loop.
+ */
+function deactivateOpenContourEdit(element: HTMLDivElement) {
+  state.isInteractingWithTool = false;
+
+  element.removeEventListener(
+    Events.MOUSE_UP,
+    this.mouseUpOpenContourEditCallback
+  );
+  element.removeEventListener(
+    Events.MOUSE_DRAG,
+    this.mouseDragOpenContourEditCallback
+  );
+  element.removeEventListener(
+    Events.MOUSE_CLICK,
+    this.mouseUpOpenContourEditCallback
+  );
+
+  resetElementCursor(element);
+}
+
+/**
+ * Adds points to the edit line and calculates the preview of the edit to render.
+ * Checks if an edit needs to be completed by crossing of lines, or by dragging
+ * the edit line past the end of the open contour.
+ */
+function mouseDragOpenContourEditCallback(
+  evt: EventTypes.MouseDragEventType
+): boolean {
+  const eventDetail = evt.detail;
+  const { currentPoints, element } = eventDetail;
+  const worldPos = currentPoints.world;
+  const canvasPos = currentPoints.canvas;
+  const enabledElement = getEnabledElement(element);
+  const { renderingEngine, viewport } = enabledElement;
+
+  const { viewportIdsToRender, xDir, yDir, spacing } = this.commonData;
+  const { editIndex, editCanvasPoints, startCrossingIndex } =
+    this.commonEditData;
+
+  const lastCanvasPoint = editCanvasPoints[editCanvasPoints.length - 1];
+  const lastWorldPoint = viewport.canvasToWorld(lastCanvasPoint);
+
+  const worldPosDiff = vec3.create();
+
+  vec3.subtract(worldPosDiff, worldPos, lastWorldPoint);
+
+  const xDist = Math.abs(vec3.dot(worldPosDiff, xDir));
+  const yDist = Math.abs(vec3.dot(worldPosDiff, yDir));
+
+  // Get pixel spacing in the direction.
+  // Check that we have moved at least one voxel in each direction.
+
+  if (xDist <= spacing[0] && yDist <= spacing[1]) {
+    // Haven't changed world point enough, don't render
+    return;
+  }
+
+  const numPointsAdded = addCanvasPointsToArray(
+    element,
+    editCanvasPoints,
+    canvasPos,
+    this.commonData
+  );
+
+  const currentEditIndex = editIndex + numPointsAdded;
```

Comment:
I first thought `editIndex` is the index of the contour being edited, but apparently it is not ...? and it is related the the number of points recently added somehow.

Reply:
So is it like you associate an index to each point being added and you use it for crossing etc? I think I see what you have done

Reply:
Its the index of the (invisible) edit line being used to generate the edit preview. Analogous to `handleIndex` of BidirectionalTool's `editData` pointing to its own handle, but for a different line used for calculation. I'll add some comments.